### PR TITLE
Feature - 3265 - Alert Dialog Component

### LINKED ISSUE #3265: Add Alert Dialog component
Based on the Alert Dialog component from [Reach](https://reach.tech/).

Source: https://reach.tech/alert-dialog

Example where this component should be used instead of the Dialog component: https://github.com/GCTC-NTGC/gc-digital-talent/blob/main/frontend/admin/src/js/components/menu/LoginOrLogout.tsx#L53:L109

### PR DESCRIPTION
**Pull request by esizer**
Resolves #3265 

## Summary

This creates a new `<AlertDialog />` component. It differs from the `<Dialog />` component in that it interrupts current user flow, announces its content to assistive technology and usually is asking for an action to be taken.

## Usage

The component has a wrapper as well as exposing the underlying components for specific customization.

### Easiest

Here, we are using the wrapper and allowing the component to construct itself.

```tsx
import AlertDialog from "@common/components/AlertDialog";

const DeleteConfirm = () => {
  const leastDestructive = React.useRef(null); // Required to draw focus to the least destructive action
  const [isOpen, setOpen] = React.useState<boolean>(false); // Controls when the alert dialog is open or closed

  return (
    <AlertDialog
      isOpen={isOpen}
      leastDestructiveRef={leastDestructive}
      onDismiss={() => setIsOpen(false)}
      title="Are you sure you'd like to delete this?"
    >
      <AlertDialog.Description>
        <p>Deleting something is permanent and cannot be undone.</p>
      </AlertDialog.Description>
      <AlertDialog.Actions>
        <button type="button" onClick={() => setIsOpen(false)} ref={leastDestructive}>Cancel</button>
        <span style={{marginLeft: "0.25rem"}}>
          <button type="button" onClick={destructiveAction}>Delete</button>
        </span>
      </AlertDialog.Actions>
    </AlertDialog>  
  );
}
```

### More Control

```tsx
import AlertDialog from "@common/components/AlertDialog";

const DeleteConfirm = () => {
  const leastDestructive = React.useRef(null); // Required to draw focus to the least destructive action
  const [isOpen, setOpen] = React.useState<boolean>(false); // Controls when the alert dialog is open or closed

  return isOpen ? (
    <AlertDialog.Overlay leastDestructiveRef={leastDestructiveRef}>
      <AlertDialog.Content>
        <AlertDialog.Heading onDismiss={() => setIsOpen(false)}>
          Are you sure you'd like to delete this?
        </AlertDialog.Heading>
        <AlertDialog.Description>
          <p>Deleting something is permanent and cannot be undone.</p>
        </AlertDialog.Description>
        <AlertDialog.Actions>
          <button type="button" onClick={() => setIsOpen(false)} ref={leastDestructive}>Cancel</button>
          <span style={{marginLeft: "0.25rem"}}>
            <button type="button" onClick={destructiveAction}>Delete</button>
          </span>
        </AlertDialog.Actions>
      </AlertDialog.Content>
    </AlertDialog,Overlay>  
  ) : null;
}
```

### CODE REVIEW

File: frontend/common/src/components/AlertDialog/AlertDialog.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import React from "react";
+import { Story, Meta } from "@storybook/react";
+import AlertDialogComponent from ".";
```
Comment (patcon) [inline frontend/common/src/components/AlertDialog/AlertDialog.stories.tsx:3]: Any reason not to import as `AlertDialog`? Seems that the story is implicit docs on how to use, so feels like a mixed msg to give a nonstandard story-specific name

File: frontend/common/src/components/AlertDialog/AlertDialog.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+              </div>
+            </AlertDialogComponent.Actions>
+          </AlertDialogComponent.Content>
+        </AlertDialogComponent.Overlay>
+      )}
+    </>
+  );
+};
+
+export const AlertDialog = Template.bind({});
```
Comment (patcon) [inline frontend/common/src/components/AlertDialog/AlertDialog.stories.tsx:62]: What about a `Basic` story (not using all the sub-components explicitly) and an `Advanced` one where all the sub-components are used individually? Again, just for the stories to be documentation of use-cases

File: frontend/common/src/components/AlertDialog/AlertDialog.stories.tsx (hunk 1 above)
Comment (esizer) [reply frontend/common/src/components/AlertDialog/AlertDialog.stories.tsx:3]: I've been messing around with implementation today to get it right before I put this up for review and I was changing names around and testing stuff out where I was using `AlertDialog` a lot so just used this name for now until I nail it down and create a proper story.

File: frontend/common/src/components/AlertDialog/AlertDialog.stories.tsx (hunk 2 above)
Comment (esizer) [reply frontend/common/src/components/AlertDialog/AlertDialog.stories.tsx:62]: Yeah, this is WIP and not the story I plan on putting up for review. I just put this up to make it easier to walk through different implementations and preview before finalizing the component.

Comment (patcon) [reply frontend/common/src/components/AlertDialog/AlertDialog.stories.tsx:62]: k i'll duck out then! Thanks!

**Comment by esizer** [comment]
@patcon Aside from the story itself being odd, what are your thoughts on how it is implemented (outlined in description)?

The goal was:

- Provide simple implementation that handles as much of the hard stuff for us
- Expose underlying components for more "control"
- Remove `footer` prop so it is easier to wrap all children in a `FormProvider`

**Comment by petertgiles** [comment]
Oh, and could you check on Chromatic?  It's not a very useful snapshot right now.

File: frontend/common/src/components/AlertDialog/AlertDialog.test.tsx [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  actions: React.ReactNode;
+} = {
+  isOpen: true,
+  onDismiss: jest.fn(),
+  title: "title",
+  children: null,
+  actions: null,
+};
+
+const TestingAlertDialog = ({ children, ...rest }: AlertDialogProps) => {
```
[the remark this answers is not in the document]
Comment (esizer) [reply frontend/common/src/components/AlertDialog/AlertDialog.test.tsx]: Added with [improve tests + add test for esc closing](https://github.com/GCTC-NTGC/gc-digital-talent/pull/3400/commits/c64a71403f2990d50d16ee6290ceccbc26bf85ac)

**Review by petertgiles — APPROVED** [review]
Love it.

**Pull request merged**